compiler/rustc_mir_build/src/build/matches/test.rs

```diff
@@ -251,6 +251,39 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
             }
 
             TestKind::Eq { value, ty } => {
+                let tcx = self.tcx;
+                if let ty::Adt(def, _) = ty.kind() && Some(def.did()) == tcx.lang_items().string() {
```

Comment:
I guess here is what you would have to look for all `Deref` impls of `def` and look for any where the types match (on both sides). What would be missing from this code is having a check, somewhere to deduplicate the resulting `.deref()` call for cases where the impl is not idempotent (and verify that the derefed value is dropped in the match arm and not at the end of the `match` block).

Reply:
(Note to self: this would also be a place to provide suggestions if no impl matches, but there are some, potentially filtered using some heuristic, maybe limited only to std.)

Reply:
I feel like you're missing a check here, even for this minimal case: you should be looking at `value.ty()` to see if it is actually `&str`, otherwise you would continue through this with, for example, a numeric literal and I don't know what the error would be in that case. Could you try out `match String::new() { 42 => {} _ => {}}` and tell us what the failure is?

Reply:
There is previous typeck code that prevents this.


Here's the output

```
error[E0308]: mismatched types
 --> a.rs:2:27
  |
2 |     match String::new() { 42 => {} _ => {}}
  |           -------------   ^^ expected struct `String`, found integer
  |           |
  |           this expression has type `String`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
```

Reply:
That makes me preemptively sad, knowing that we'll have to have deref checks in E0308 both ways to account for deref in patterns. :(

Reply:
It could just be a `TestKind` variant encoding the necessary deref steps so we don't have to check again